### PR TITLE
don't fail on introspection queries

### PR DESCRIPTION
**Pull request by chirag04**
I haven't made this configurable yet. open to thoughts.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -136,10 +137,15 @@ export function createComplexityLimitRule(maxCost, options = {}) {
 
     return {
       enter(node) {
+        // skip introspection queries
+        if (node.kind === 'Field' && node.name.value === '__schema') {
```
Comment (jquense) [inline src/index.js]: really any field with `__` though right? internal fields like `__typename` are free i think

**Comment by chirag04** [comment]
`this isn't right; the BREAK return would disable all cost checking on any queries that looked at the schema` that's exactly what i'm trying here.

`want to avoid potential problems with letting people run arbitrary queries against __schema anyway`. these queries are for free. so y bother?

**Comment by chirag04** [comment]
ok. wait. i see what you mean. We should check for the operation. wondering if graphql allows multiple query operations tho.

**Comment by chirag04** [comment]
ignoring private fields(__) and returning false should address the problem here.

**Comment by taion** [comment]
what does "free" mean?

if someone sets up a query hitting only "free" fields that still e.g. requires multiple gigabytes of response size, then that would still kill the server

so disabling all cost checking there just doesn't seem like the right thing. this patch would allow e.g.

```
{ list { list { list { list { list { list { list { list { list { __typename ...
```

and that's a server-killer. and that's bad.

**Comment by taion** [comment]
well, does it work?

**Comment by taion** [comment]
is this the right level to do this, btw? perhaps the caller should special-case the introspection query instead?

**Comment by chirag04** [comment]
> well, does it work?

It does. y not? Added a test case as well.

> is this the right level to do this, btw? perhaps the caller should special-case the introspection query instead?

who is the `caller` here?

**Comment by taion** [comment]
I mean the external thing setting up the rules in the first place.

Have you looked into why the introspection query is so expensive yet?

**Comment by chirag04** [comment]
> I mean the external thing setting up the rules in the first place.

I think this is the right place to do it because every caller will have this problem. We can put it behind a flag if required.

> Have you looked into why the introspection query is so expensive yet?
Total cost: `28553` out of which FullType fragment has a cost of `26653`. it's full of nested Lists. 
https://github.com/graphql/graphql-js/blob/master/src/utilities/introspectionQuery.js#L34-L65
https://github.com/graphql/graphql-js/blob/master/src/type/introspection.js#L260-L300

**Comment by jquense** [comment]
seems like this is the right place to do this...why would ever be interested in the cost of the introspection query, its a constant cost regardless of your schema

**Comment by taion** [comment]
Sure but there are multiple possible introspection queries... it's not guaranteed to never change

**Comment by chirag04** [comment]
> Sure but there are multiple possible introspection queries... it's not guaranteed to never change

I'm mostly interested in this specific introspection query from graphql. everything else is a legitimate attack vector and should go through cost calculation?

**Comment by taion** [comment]
There are variants: https://github.com/graphql/graphiql/blob/v0.11.1/src/utility/introspectionQueries.js#L11-L105

Also it changes around:https://github.com/graphql/graphql-js/commit/0b72e7038761bec9fb5319cc08ea93ff8b071a9c#diff-4381e15d00ce4b749116b59bcbee9863

**Comment by chirag04** [comment]
i'm not worried about the changes to introspection query. we are importing from graphql package. I did see that variant and that's legit. maybe import the introspection query or the variant used by graph**I**ql?

**Comment by taion** [comment]
I really don't think this is the way to go. Among other things, the deep equality check here is not cheap:

```
    ✓ should not fail on introspection queries (526ms)
```

Per #7, it's relatively easy to just discount introspection list queries to reduce the cost to below 1000. By contrast, this doesn't blow out the runtime at all:

```
    ✓ should not fail on default introspection query with limit 1000 (25ms)
```

**Comment by chirag04** [comment]
feel free to close this in favor of #7

**Comment by taion** [comment]
I re-ran timing and it looks even worse:

```
    ✓ should not fail on introspection queries (987ms)
```

Replacing with #7

**Comment by chirag04** [comment]
weird that it consistently takes about 300ms on my machine. anyway, lets ship #7.

`    ✓ should not fail on introspection queries (331ms)`

**Comment by taion** [comment]
That's still a really long time to tie up the CPU on a Node worker.

**Pull request closed without merging**